Extract metadata methods from IAppResources to separate interface and add feature flags section (resolves #188)

tjololo opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
Application metadata response extends Application model from Altinn.Platform.Storage.Interface.Models and adds list of supported and enabled features necessary for frontend to support multiple backends

## Related Issue(s)
Fixes #188 

## Verification
- [ ] **Your** code builds clean without any errors or warnings
- [x] Manual testing done (required)
- [x] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [x] All tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

tjololo committed:
Implement features part of application metadata.

Move app metadata methods behind separate interface IAppMetadata.cs
Clean up some code
Mark moved methods as Obsolete in IAppResources.cs

tjololo committed:
Added tests for IAppMetadata.cs

tjololo committed:
Removed last usages of IAppResources.GetApplication

tjololo committed:
AppResourcesSI.cs fetches metadata files from new Interface IAppMetdata.

Added tests to verify result equals old result

tjololo committed:
Fix test errors due to operation systems not agreeing on how to do a linebreak

tjololo committed:
Fix some warnings from sonar

tjololo committed:
Fix bug in PersonClient.cs

tjololo committed:
More codesmell fixes

tjololo committed:
Fix some more sonar warnings

tjololo committed:
Fix more nullability warnings

tjololo commented:
Alot of nullability checks due to the fact that if read of applicationmetadata.json fails we return null.
Have created an issue #202 for this

ivarne commented on src/Altinn.App.Core/Infrastructure/Clients/Register/PersonClient.cs:
```diff
@@ -64,16 +63,17 @@ public class PersonClient : IPersonRetriever
-
-            await AddAuthHeaders(request);
+            using(var request = new HttpRequestMessage(HttpMethod.Get, $"persons"))
+            {
```
There is now new syntax for a using declaration that creates less indentation https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/using-statement
```suggestion
            using var request = new HttpRequestMessage(HttpMethod.Get, $"persons");
```

Just skip the `(` after using and end the line with a `;`

tjololo replied:
Nice! Thanks, great readability improvement!

ivarne replied:
UUps, Should be sure to create a full example when using the suggest feature.

tjololo committed:
Update src/Altinn.App.Core/Infrastructure/Clients/Register/PersonClient.cs

Co-authored-by: Ivar Nesje <ivarne@users.noreply.github.com>

tjololo committed:
Remove closing brace

tjololo commented:
> Alot of nullability checks due to the fact that if read of applicationmetadata.json fails we return null. Have created an issue #202 for this

I have tested renaming and supplying invalid applicationmetdata.json files in an application. 
Missing file result in crash at startup. Missconfigures (eg: string in an int field) result in crash when instance is created. 
I think we can safely remove the nullable from GetApplicationMetdata and have it throw and Exception instead. 
This would remove the null checks that only occurs when the application crashes either way.

What do you think @RonnyB71 ?

tjololo committed:
Make AppMedataTest.cs throw exception if applicationmetadata.json is missing or invalid.

Replaced newtonsoft with System.Text.Json
Added test for common config for eFormidling
Removed nullchecks for ApplicationMetadata as it newer should be null.
TODO: Decide if the same changes (exception instead of null) for failure to deserialize

tjololo committed:
Reference Interface instead of implementation in obsolete message

tjololo commented on src/Altinn.App.Core/Implementation/AppResourcesSI.cs:
```diff
@@ -104,66 +109,49 @@ public byte[] GetText(string org, string app, string textResource)
+                return null;   
             }
+            _logger.LogError("Something went wrong fetching application metadata");
+            return null;
```
@RonnyB71 should we just ignore these exceptions here as well? Would change how the class works but we are no longer returning null on a non-nullable return type

RonnyB71 replied:
Yes, I think so. Does not make sense to swallow the exception here and let it blow up somewhere else.

RonnyB71 replied:
And let's create version 8 just to signal that we are changing behavior even though it would crash in a different place anyway.

tjololo replied:
throwing ApplicationConfigException, the interface is promising a value so I think we don't need to update the major version. Either way the application will crash

RonnyB71 commented on src/Altinn.App.Core/Internal/App/IAppFeatures.cs:
```diff
@@ -0,0 +1,14 @@
+        /// Fetch frontend features that are supported by this backend
+        /// </summary>
+        /// <returns>List of frontend features enabled/disabled for this backend</returns>
+        public Task<Dictionary<string, bool>> GetEnabledFeatures();
```
Does it get enabled features or enabled/disabled features - perhaps rename to GetFeatures?

tjololo replied:
Agreed, will change it as soon as we conclude if we should rename features to capabilities

RonnyB71 commented on src/Altinn.App.Core/Internal/App/AppMetadata.cs:
```diff
@@ -0,0 +1,118 @@
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Something went wrong when fetching BPMNProcess");
```
Should we do the same for BPMN as for ApplicationMetadata? The app will crash without it so might as well let it crash and burn.

ivarne replied:
We should do an app validation on startup to check everything as I suggested in https://github.com/Altinn/app-lib-dotnet/pull/98

tjololo replied:
I totally agree we should add some sort of startup validation. If we could make it easy to extend the startup validation we could tackle the files one at a time and also give each app developer the possibility to extend what checks are triggered at startup.

RonnyB71 replied:
Sounds  good. But I guess we would like to get this out of the door and not add any more changes? Still good to just throw the exception where it happens.

tjololo replied:
Created an issue to track validation on startup #208 to not make this PR even larger

RonnyB71 commented on src/Altinn.App.Core/Internal/App/AppFeatures.cs:
```diff
@@ -0,0 +1,24 @@
+    /// <summary>
+    /// Default implementation of IAppFeatures
+    /// </summary>
+    public class AppFeatures : IAppFeatures
```
We should perhaps document the difference between this feature handling and the FeatureManagement use somewhere in the code as it's there it might be confusing for others (and future me).

tjololo replied:
Should we consider renaming our implementation to capabilities to make the likelihood of confusing it with FeatureManagement?
@olemartinorg if we decide to rename this, does it seem like a better name in appmetadata? we are reporting if the backend has the capability X not a feature?

olemartinorg replied:
Fine by me - but we'd need to change it in the next release of app-frontend in that case (the multipart-save feature/capability is already in production).

Let's consider the two cases we have now:
- Footer. This is a new capability the backend magically gets in a new version.
- Multipart saving (should probably be named "changelist in data write" or something on the backend), which is a capability you should be able to turn on/off in apps (and which is on by default for new apps, and will be the only option in some future major version).

From that, I think 'features' sound better, because features can be toggled and capabilities sounds like they cannot be toggled. The difference is that `FeatureManagement` lets an app developer control the backend/frontend features of their app, while `AppFeatures` lets the backend toggle features in app-frontend. So maybe `FrontendFeatures` is a better name? 🤷

tjololo replied:
Agreed, was only thinking of footer when I came up with capabilities. FrontendFeatures sound better as our intent is to report what frontend features the backend supports and has enabled. Any objections to FronendFeatures @RonnyB71 ?

RonnyB71 replied:
Yes. I would like it to be FrontendFeatures ;-)

tjololo replied:
Renamed it to FrontendFeatures

RonnyB71 commented on src/Altinn.App.Core/Implementation/AppResourcesSI.cs:
```diff
@@ -104,66 +109,49 @@ public byte[] GetText(string org, string app, string textResource)
-                _logger.LogError("Something went wrong when fetching application metadata. {0}", ex);
-                return null;
+                _logger.LogError(ex, "Something went wrong fetching application metadata");
+                return null;   
```
Let it crash and burn?

tjololo replied:
Crash and burn implemented in latest version :)

RonnyB71 commented on src/Altinn.App.Core/Models/ApplicationMetadata.cs:
```diff
@@ -0,0 +1,24 @@
+        /// <summary>
+        /// Name of the application
+        /// </summary>
+        public string App { get; set; } = "";
```
Should we consider using AppIdentifier instead of continuing to require both app and org in signatures or taking the whole ApplicationMetadata object when you only needs those two in combination?

tjololo replied:
Made some changes and added AppIdentifier instead of string App. Override set of Id to ensure AppIdentifier is updated if Id is updated

RonnyB71 commented on src/Altinn.App.Api/Controllers/InstancesController.cs:
```diff
@@ -484,10 +475,10 @@ public class InstancesController : ControllerBase
         }
 
-        private async Task CopyDataFromSourceInstance(Application application, Instance targetInstance, Instance sourceInstance)
+        private async Task CopyDataFromSourceInstance(ApplicationMetadata application, Instance targetInstance, Instance sourceInstance)
```
An example where I think a dependency on ApplicationMetadata is too broad since we only need the app/org or AppIdentifier.

tjololo replied:
Agreed, I belive this is a weakness in the Application model, we have and Id containing org and appname (org/appname) but also a separate field for org. These could potentially become out of sync and return different orgs.
I have a suggestion for adding AppIdentifier to ApplicationMetadata class instead of string App

tjololo committed:
Rename after review

* AppFeatures to FrontendFeatures
* GetEnabledFeatures to GetFeatures

tjololo committed:
Throw ApplicationConfigException if metdata or BPMN is invalid.
IAppResources.GetMetadata changed to throw ApplicationConfigException as well

tjololo committed:
document logger

tjololo committed:
Add Serializable attribute

RonnyB71 commented on src/Altinn.App.Core/Internal/App/AppMetadata.cs:
```diff
@@ -0,0 +1,118 @@
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Something went wrong when fetching XACML Policy");
+            }
```
Perhaps add a comment why we allow swallowing this exception and include the file path in the  message as well.

RonnyB71 commented on src/Altinn.App.Core/Internal/App/AppMetadata.cs:
```diff
@@ -0,0 +1,112 @@
+            {
+                if (File.Exists(filename))
+                {
+                    JsonSerializerOptions jsonSerializerOptions = new JsonSerializerOptions()
```
Ideally we should have these as default settings thus avoiding having to repeat them, but I'm not sure if it's possible. Just something we could keep in mind if we end up repeating these all over.

tjololo replied:
Agreed, have removed the generic exception part as I didn't affect anything authorisation wise

RonnyB71 commented on src/Altinn.App.Core/Internal/App/AppMetadata.cs:
```diff
@@ -0,0 +1,112 @@
+                return _application;
+            }
+
+            string filename = _settings.AppBasePath + _settings.ConfigurationFolder + _settings.ApplicationMetadataFileName;
```
Not you code, but a Path.Join would be good here.

tjololo replied:
Nice catch, replacing all similar lines in AppMetadata with Path.Join

RonnyB71 commented on src/Altinn.App.Core/Implementation/AppResourcesSI.cs:
```diff
@@ -104,67 +109,52 @@ public byte[] GetText(string org, string app, string textResource)
-                _logger.LogError("Something went wrong when fetching XACML Policy. {0}", ex);
+                return task.Result;
             }
 
```
Same here, worth a comment on why we allow swallowing this particular exception while the others are thrown.

tjololo committed:
Fix comments after review

tjololo committed:
Add AppIdentifier instead of App to ApplicationMetadata.cs field is omitted from Json but available in code.

Ensure AppIdentifier is in sync with Id

tjololo committed:
Fix some code smells

tjololo committed:
Add test with invalid app id. Refactor set to make logic more readable
